src/core/vectortile/qgsvectortilebasicrenderer.cpp

```diff
@@ -162,14 +162,42 @@ void QgsVectorTileBasicRenderer::stopRender( QgsRenderContext &context )
   Q_UNUSED( context )
 }
 
+void QgsVectorTileBasicRenderer::renderBackground( QgsRenderContext &context )
+{
+  for ( const QgsVectorTileBasicRendererStyle &layerStyle : std::as_const( mStyles ) )
+  {
+    if ( !layerStyle.symbol() || layerStyle.layerName() != QLatin1String( "background" ) )
+      continue;
+
+    QgsExpression filterExpression( layerStyle.filterExpression() );
+    filterExpression.prepare( &context.expressionContext() );
+
+    QgsSymbol *sym = layerStyle.symbol();
+    sym->startRender( context, QgsFields() );
+
+    QgsFillSymbol *fillSym = dynamic_cast<QgsFillSymbol *>( sym );
+    if ( fillSym )
+    {
+      QPolygon polygon;
+      polygon << QPoint( 0, 0 );
+      polygon << QPoint( 0, context.outputSize().height() );
+      polygon << QPoint( context.outputSize().width(), context.outputSize().height() );
+      polygon << QPoint( context.outputSize().width(), 0 );
+      fillSym->renderPolygon( polygon, nullptr, nullptr, context );
```

Comment:
polygon should be closed, right?

Reply:
@nyalldawson , I don't think so, it wasn't closed before (i.e. when we were relying on the tilePolygon() which was a 4-point polygon created by the QgsVectorTileUtils::tilePolygon( ... ) function.